Source/WebCore/dom/Element.cpp

```diff
@@ -2735,232 +2735,306 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
-            setHasDirAutoFlagRecursively(this, false);
-        if (auto* input = dynamicDowncast<HTMLInputElement>(*this); parent && parent->usesEffectiveTextDirection() && !(input && input->isTelephoneField()))
-            updateEffectiveDirectionality(parent->effectiveTextDirection());
-        else
-            updateEffectiveDirectionality(std::nullopt);
-        break;
-    case TextDirectionDirective::LTR:
-        if (selfOrPrecedingNodesAffectDirAuto())
-            setHasDirAutoFlagRecursively(this, false);
-        updateEffectiveDirectionality(TextDirection::LTR);
-        break;
-    case TextDirectionDirective::RTL:
-        if (selfOrPrecedingNodesAffectDirAuto())
-            setHasDirAutoFlagRecursively(this, false);
-        updateEffectiveDirectionality(TextDirection::RTL);
-        break;
-    case TextDirectionDirective::Auto:
-        setUsesEffectiveTextDirection(true);
-        updateEffectiveDirectionalityOfDirAuto();
-        break;
+    // Specs: https://html.spec.whatwg.org/multipage/dom.html#attr-dir
+    switch (textDirectionState) {
+    case TextDirectionState::LTR:
+        return TextDirection::LTR;
+    case TextDirectionState::RTL:
+        return TextDirection::RTL;
+
+    case TextDirectionState::Auto:
+        // Specs: Return the auto directionality of element.
+        // Specs: If directionality is null, then return 'ltr'.
+        return element.computeAutoTextDirection().value_or(TextDirection::LTR);
+
+    case TextDirectionState::Undefined:
+        // Specs: If element is a bdi element, then return the auto directionality of element.
+        // Specs: If directionality is null, then return 'ltr'.
+        if (is<HTMLBDIElement>(element))
+            return element.computeAutoTextDirection().value_or(TextDirection::LTR);
+
+        // Specs: If element is an input element whose type attribute is in the Telephone state
+        // then return LTR;
+        if (auto* input = dynamicDowncast<HTMLInputElement>(element); input && input->isTelephoneField())
+            return TextDirection::LTR;
+
+        // Specs: If the parent is a shadow root, then return the directionality of parent's host.
+        // Specs: If the parent is an element, then return the directionality of parent.
+        if (RefPtr parent = element.parentOrShadowHostElement())
+            return computeElementTextDirection(*parent, elementTextDirectionState(*parent));
+
+        return std::nullopt;
     }
 
-    if (parent && parent->selfOrPrecedingNodesAffectDirAuto()) {
-        if (isValid && direction != TextDirectionDirective::Auto)
-            setHasDirAutoFlagRecursively(this, false);
-        parent->adjustDirectionalityIfNeededAfterChildAttributeChanged(this);
+    return std::nullopt;
+}
+
+static std::optional<TextDirection> computeElementContainedTextAutoDirection(const Element& element)
```

Comment:
Can we add a spec URL here?

Reply:
This function name is also super confusing. Maybe spec has some name we can adopt here?

Reply:
This is https://html.spec.whatwg.org/#contained-text-auto-directionality, right? We should just call it computeContainedTextAutoDirectionality and add a link here.

Reply:
I got this name from the specs: https://html.spec.whatwg.org/multipage/dom.html#contained-text-auto-directionality.

Reply:
I don't think the extra "element" is adding any value here.